### PR TITLE
Single match else

### LINKED ISSUE #172: single_match suggestion should handle else blocks
```
match x {
 something => {foo}
 _ => {bar}
}
```

gets the suggestion

```
if let something = x {
 foo
}
```

but it should get the suggestion

```
if let something = x {
 foo
} else {
 bar
}
```

Just needs some extra snippets in https://github.com/Manishearth/rust-clippy/blob/master/src/misc.rs#L50 if there is an else block

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #172

### CODE REVIEW

File: src/matches.rs [hunk 1]
```diff
@@ -112,34 +112,40 @@ impl LateLintPass for MatchPass {
 fn check_single_match(cx: &LateContext, ex: &Expr, arms: &[Arm], expr: &Expr) {
     if arms.len() == 2 &&
        arms[0].pats.len() == 1 && arms[0].guard.is_none() &&
-       arms[1].pats.len() == 1 && arms[1].guard.is_none() &&
-       is_unit_expr(&arms[1].body) {
+       arms[1].pats.len() == 1 && arms[1].guard.is_none() {
```
Comment (Manishearth) [inline src/matches.rs]: Perhaps we should have an `Allow` lint for matches with an else block? I don't think one style is clearly preferred over the other, cc @llogiq .

Comment (llogiq) [inline src/matches.rs]: Yeah, I'm also not sure about this – on one hand the `match` statement has one more level of nesting, which may lead to more line breaks which hurt readability, on the other hand `if let` will often lead to one more line of *very* sparse code, which takes more vertical space then necessary.

Comment (oli-obk) [inline src/matches.rs]: The conclusion is: add a lint that suggests turning regular `if/else` into a match on a boolean xD

One thing I have noticed is that I find statements much more readable with `if let` and expressions with `match`

Comment (oli-obk) [inline src/matches.rs]: if the match arm is a (multiline) block, the `if let` version takes up one line less:

```rust
match x {
    Y => println!("Jupiter"),
    _ => {
        println!("Mars");
        println!("Venus");
    },
}
```

vs

```rust

if let Y = x {
    println!("Jupiter");
} else {
    println!("Mars");
    println!("Venus");
}

Comment (llogiq) [inline src/matches.rs]: True. But I'm more worried about

```
if let Some(x) = whaa {
    x
} else {
    0
}
```

which would obviously better be written as `whaa.unwrap_or(0)`...

Comment (oli-obk) [inline src/matches.rs]: I can change the check to only show up if there's no else arm or if any of the arms is a block with at least a statement and an expression

Comment (Manishearth) [inline src/matches.rs]: SGTM

**Comment by Manishearth** [comment]
On the bikeshed issue someone pointed out that this is a matter of personal style. I think we should just unconditionally allow (as in, a pedantic lint) matches with an else block.

**Comment by oli-obk** [comment]
I have added a new lint and made it allow by default. The new lint is also much better than before. There is now no (imo) uncontroversial fallout in clippy.

**Pull request merged**